Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx

```diff
@@ -0,0 +1,602 @@
+                int deId, int padId, int adc, Digit::Time time, Digit::Time stopTime);
+
+  void mergeDigits(int feeId);
+};
+
+void FeeIdMerger::setOrbit(uint32_t orbit, bool stop)
+{
+  // send digits of previous orbit if either the stop RDH is received
+  // or a new orbit is started
+  if ((orbit == buffers[currentBufId].orbit) && (!stop)) {
+    return;
+  }
+
+  int nSent = 0;
+  for (auto& d : buffers[previousBufId].digits) {
+    if (!d.merged && (d.digit.getPadID() >= 0)) {
+      uint16_t size = (d.stopTime.getBXTime() - d.digit.getTime().getBXTime()) / 4 + 1;
+      sendDigit(d.digit);
+      nSent += 1;
+    }
+  }
+  if (mPrint) {
+    std::cout << "[FeeIdMerger] sent " << nSent << " digits for orbit " << buffers[previousBufId].orbit << "  current orbit is " << orbit << std::endl;
+  }
+
+  currentBufId = 1 - currentBufId;
+  previousBufId = 1 - previousBufId;
+  buffers[currentBufId].digits.clear();
+  buffers[currentBufId].orbit = orbit;
+}
+
+void FeeIdMerger::mergeDigits()
+{
+  auto checkDigits = [](MergerDigit& d1, MergerDigit& d2) -> bool {
+    // skip digits that are already merged
+    if (d1.merged || d2.merged)
+      return false;
+
+    // skip all digits not matching the detId/padId
+    if (d1.solarId != d2.solarId)
+      return false;
+    if (d1.dsAddr != d2.dsAddr)
+      return false;
+    if (d1.chAddr != d2.chAddr)
+      return false;
+    if (d1.digit.getDetID() != d2.digit.getDetID())
+      return false;
+    if (d1.digit.getPadID() != d2.digit.getPadID())
+      return false;
+
+    // compute time difference
+    Digit::Time startTime = d1.digit.getTime();
+    uint32_t bxStart = startTime.bunchCrossing;
+    Digit::Time stopTime = d2.stopTime;
+    uint32_t bxStop = stopTime.bunchCrossing;
+    // correct for value rollover
+    if (bxStart < bxStop)
+      bxStart += 0x100000;
+
+    uint32_t stopTimeFull = bxStop + (stopTime.sampaTime << 2);
+    uint32_t startTimeFull = bxStart + (startTime.sampaTime << 2);
+    uint32_t timeDiff = startTimeFull - stopTimeFull;
+
+    if (mPrint) {
+      std::cout << "checkDigits: " << d1.digit.getDetID() << "  " << d1.digit.getPadID() << "  " << bxStop << "  " << stopTime.sampaTime
+                << "  " << bxStart << "  " << startTime.sampaTime << "  " << timeDiff << std::endl;
+    }
+
+    // skip if the time difference is not equal to 1 ADC clock cycle
+    if (timeDiff > 8)
+      return false;
+
+    // merge digits
+    d2.digit.setADC(d1.digit.getADC() + d2.digit.getADC());
+    d2.stopTime = d1.stopTime;
+    d1.merged = true;
+    return true;
+  };
+
+  auto& currentBuffer = getCurrentBuffer();
+  auto currentBufId = getCurrentBufId();
+  auto& previousBuffer = getPreviousBuffer();
+  auto previousBufId = getPreviousBufId();
+
+  if (mPrint) {
+    std::cout << "Merging digits in " << previousBufId << " (orbit=" << previousBuffer.orbit << ")\n";
+  }
+
+  for (size_t i = 0; i < previousBuffer.digits.size(); i++) {
```

Comment:
I can't say I understand why this loop is there. The second one below which merges current with previous I think I get, but not this one... Care to explain ?

Reply:
Sure... let me try to briefly outline the whole data flow. If the explanation is clear enough we might also put it as documentation in the source file.

First of all, during the processing of orbit N the corresponding digits are added to the current merge buffer (there is one merge buffer for each FEEID).

When the end of orbit N is reached (either because we receive the stop RDH or because the orbit number increases), the merging is performed. First, the not-yet-merged digits in the previous buffer are merged together, then the merging between the previous and current buffers is performed. The "seeds" of the merging algorithm are digits for which the first ADC sample is exactly at the beginning of the SAMPA time window, i.e. the SAMPA "cluster time" is equal to zero. Since we are only merging digits belonging to the same SAMPA channel, and that the data coming from each SAMPA channel is already in time-ascending order, a single loop on the seeds followed by an inner loop on the non-seed digits is enough to catch all combinations.

When the merging of the two buffers is completed, the merged digits of the previous buffer are sent out. There is therefore an offset of one orbit between the raw decoding and the digits output. This in particular means that the digits belonging to the last orbit of a given input buffer (typically a TimeFrame) will not be sent to the output, because they would need to be merged with the those from the first orbit of the next TF.

Does it clarify your doubts?